teps/0121-refine-retries-for-taskruns-and-customruns.md

```diff
@@ -0,0 +1,200 @@
+---
+status: proposed
+title: Refine Retries for TaskRuns and CustomRuns
+creation-date: '2022-09-08'
+last-updated: '2022-10-05'
+authors:
+- '@XinruZhang'
+---
+
+# TEP-0121: Refine Retries for TaskRuns and CustomRuns
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Notes and Caveats](#notes-and-caveats)
+- [Design Details](#design-details)
+- [Alternatives](#alternatives)
+- [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Two distinct imperfections on `Retries` we'd like to address in this TEP:
+- `Retries` on `Timeout` is designed inconsistently between TaskRun and CustomRun.
+  - For CustomRun, [the document](https://github.com/tektoncd/community/blob/33ca1d5254a405b1d479f2350443f6c7979a0b72/teps/0069-support-retries-for-custom-task-in-a-pipeline.md#proposal) instructs developers to **set `Timeout` for all retry attempts**. While the actual implementation is slightly different. See the [ref](https://github.com/tektoncd/pipeline/issues/5582).
+  - For TaskRun, the `Timeout` is **set for each retry attempt**. 
+- Both `PipelineRun` reconciler and `TaskRun`|`CustomRun` reconciler are partially responsible for implementing the `Retries` as of today. See https://github.com/tektoncd/pipeline/issues/5248.
+
+To list details of the current implementation:
+
+| Runtime Object | Retriable Independently? | Retriable within a Pipeline? | Timeout for Each Attempt? | Note |
+|:---|:---|:---|:---|:---|
+|**TaskRun**| No | Yes | Yes | Implemented in PipelineRun reconciler by **updating `retriesStatus`**|
+|**CustomRun**| Yes| Yes | Yes and No | Should be implemented in CustomRun reconciler. PipelineRun controller uses `retriesStatus` to determine if the `CustomRun` finishes.|
+
+### Goals
+1. Align the behavior of `Retries` when timed out between TaskRun and CustomRun.
+2. `Retries` implementation separation.
+
+### Non-Goals
+1. Define retries behavior for PipelineRuns.
+
+### Use Cases
+#### Retry when Timeout
+
+The behavior alignment improves UX. Considering the following example:
+```
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: custom-task-pipeline
+spec:
+  tasks:
+  - name: task-run-example
+    taskRef:
+      name: task-run-example
+    retries: 1
+    timeout: "10s"
+  - name: custom-run-example
+    taskRef:
+      apiVersion: example.dev/v1alpha1
+      kind: Example
+    retries: 1
+    timeout: "10s"
+```
+
+Say customers define two child resources within a PipelineRun:
+- `task-run-example`
+- `custom-run-example`
+
+They set both `retries` and `timeout` for the two resources, under the current implementation, the two runtime objects would behave differently, which is not intuitive.
+- `task-run-example` will be retried once after 10s.
+- `custom-run-example` will be timed out after 10s.
```

Comment:
may be useful to note that if they implement retries for *each* attempt, like in taskruns instead of the documented guidance, then the `custom-run-example` would work similarly to the `custom-run-example`

Reply:
SG, added!